fhir-database-utils/src/main/java/com/ibm/fhir/database/utils/query/expression/StringExpNodeVisitor.java

```diff
@@ -0,0 +1,406 @@
+/*
+ * (C) Copyright IBM Corp. 2021
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.ibm.fhir.database.utils.query.expression;
+
+import java.math.BigDecimal;
+import java.time.Instant;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.ibm.fhir.database.utils.api.IDatabaseTranslator;
+import com.ibm.fhir.database.utils.query.Select;
+import com.ibm.fhir.database.utils.query.node.BigDecimalBindMarkerNode;
+import com.ibm.fhir.database.utils.query.node.BindMarkerNode;
+import com.ibm.fhir.database.utils.query.node.DoubleBindMarkerNode;
+import com.ibm.fhir.database.utils.query.node.ExpNode;
+import com.ibm.fhir.database.utils.query.node.ExpNodeVisitor;
+import com.ibm.fhir.database.utils.query.node.InstantBindMarkerNode;
+import com.ibm.fhir.database.utils.query.node.IntegerBindMarkerNode;
+import com.ibm.fhir.database.utils.query.node.LongBindMarkerNode;
+import com.ibm.fhir.database.utils.query.node.StringBindMarkerNode;
+
+
+/**
+ * Renders the expression node tree into a string
+ */
+public class StringExpNodeVisitor implements ExpNodeVisitor<String> {
+    private final List<BindMarkerNode> bindMarkers;
+
+    // Pretty-print the SQL statement for improved readability
+    private final boolean pretty;
+
+    public static final String NEWLINE = System.lineSeparator();
+
+    private final IDatabaseTranslator translator;
+
+    /**
+     * Simple rendering of the expression tree to a string, ignoring
+     * the bind marker values
+     */
+    public StringExpNodeVisitor() {
+        this.translator = null;
+        this.bindMarkers = null;
+        this.pretty = false;
+    }
+
+    public StringExpNodeVisitor(boolean pretty) {
+        this.translator = null;
+        this.bindMarkers = null;
+        this.pretty = pretty;
+    }
+
+    /**
+     * Collect the bind marker values into the given list
+     * @param collectBindMarkersInto
+     */
+    public StringExpNodeVisitor(IDatabaseTranslator translator, List<BindMarkerNode> collectBindMarkersInto, boolean pretty) {
+        this.translator = translator;
+        this.bindMarkers = collectBindMarkersInto;
+        this.pretty = pretty;
+    }
+
+    /**
+     * Helper to render an expression as a string using this class as a visitor
+     * @param exp
+     * @return
+     */
+    public static String stringify(ExpNode exp) {
+        return exp.visit(new StringExpNodeVisitor(true));
+    }
+
+    @Override
+    public String paren(String expr) {
+        StringBuilder result = new StringBuilder();
+        result.append("(");
+        result.append(expr);
+        result.append(")");
+        return result.toString();
+    }
+
+    @Override
+    public String and(String left, String right) {
+        StringBuilder result = new StringBuilder();
+        result.append(left);
+        if (pretty) {
+            result.append(NEWLINE).append("        ");
```

Comment:
maybe insert this as a constant  - it's reused in a couple of spots

Reply:
different number of spaces each time, to make things pretty.